### ISSUE TITLE
Impossible to annotate false positives as differing versions produce different errrors

### ISSUE THREAD

**Issue by markshannon** [python/mypy#18651]
For the following code:
```
if x:
    y = 1
if x:
    return y
```
a recent version of mypy will produce a false positive: `error: Name "y" may be undefined`.
It is not the false positive that is the issue here, but this:

If I change the code to:
```
if x:
    y = 1
if x:
    return y # type: ignore
```
then mypy 0.94 complains that `error: Unused "type: ignore" comment`.

* Why is a `type: ignore` treated as an error in the first place? That seems contrary to the normal meaning of "ignore".
* How am I to annotate code so that it works across type checkers with differing levels of false positives?

**Comment by TeamSpen210** [comment]
The `unused-ignore` error code actually isn’t [enabled by default](https://mypy.readthedocs.io/en/stable/error_code_list2.html#check-that-type-ignore-comment-is-used-unused-ignore). An ignore could mask some other issue on the same line in the future, so can be useful to let you find and remove unnecessary ones. It’s also useful in test cases, to verify a checker does error in expected locations.

To suppress the error, you need to explicitly specify the error code with mypy’s syntax, like `# type: ignore[unused-ignore, some-other-code]`. Additionally you can do `# type: ignore[…] # pyright: ignore[…]` to ignore different codes for mypy and pyright.

**Comment by markshannon** [comment]
> The unused-ignore error code actually isn’t enabled by default

It is enabled with `--strict` though, at least it seems to be.

> To suppress the error, you need to explicitly specify the error code with mypy’s syntax ...

Why make it so difficult? If the comment says "type: ignore" then do that.

**Comment by markshannon** [comment]
Apparently, more recent versions of mypy now refuse to ignore errors.

https://github.com/python/cpython/actions/runs/13242848572/job/36962057885?pr=129953
`Tools/cases_generator/stack.py:734: error: "type: ignore" comment without error`

Is mypy supposed to be this awkward or is this a bug?

**Comment by A5rocks** [comment]
By default mypy allows `type: ignore`. Sometimes you want to be more fine-grained about the error you catch!

1. Mypy has an option to tell people to be more fine grained
2. Mypy has an option to tell people that their type ignore doesn't actually ignore anything (and should be removed)

Sometimes mypy will warn differently based on the environment. Then, if for instance you have the option in 2 enabled, you need to tell mypy it's if the ignore is unused. So in your example, like mentioned above, `# type: ignore[unused-ignore, the other error code]`. (Obviously 2 shouldn't be silenced by a broad `# type: ignore` because then it wouldn't be useful without 1!)
